theano/gpuarray/opt.py

```diff
@@ -215,6 +247,224 @@ def apply(self, fgraph):
+
+    def add_requirements(self, fgraph):
+        fgraph.attach_feature(toolbox.ReplaceValidate())
+
+    def apply(self, fgraph):
+        mapping = {}
+        time_opts = {}
+        node_created = {}
+        process_count = {}
+        t_topo = time.time()
+        topo = fgraph.toposort()
+        time_topo = time.time()
+        toposort_timing = time_topo - t_topo
+
+        # Building a new graph
+        # Iterating through inputs of graph
+        for i in fgraph.inputs:
+            if isinstance(i.type, tensor.TensorType):
+                mapping[i] = as_gpuarray_variable(i, None)  # TODO context
+            else:
+                mapping[i] = i
+        for i in fgraph.variables:
+            if isinstance(i, theano.Constant):
+                mapping[i] = i
+        for node in topo:
+            for lopt in (self.local_optimizers_map.get(node.op, []) +
+                         self.local_optimizers_map.get(type(node.op), []) +
+                         self.local_optimizers_all):
+                    process_count.setdefault(lopt, 0)
+                    time_opts.setdefault(lopt, 0)
+                    node_created.setdefault(lopt, 0)
+
+        for node in topo:
+
+            if isinstance(node.op, HostFromGpu):
+                mapping[node.outputs[0]] = node.inputs[0]
+                continue
+
+            # Move only if any of the inputs are on the GPU.
+            move_to_GPU = True
+
+            '''
+            import GpuArrayVariable and GpuArraySharedVariable when you uncomment
+            if any([isinstance(i, GpuArrayVariable) or
+                   isinstance(i, GpuArraySharedVariable)
+                   for i in [mapping[v] for v in node.inputs] +
+                   node.outputs]):
+
+                move_to_GPU = True
+            '''
+
+            context_name = None
+            for i in [mapping[i] for i in node.inputs]:
+                if isinstance(i.type, GpuArrayType):
+                    context_name = i.type.context_name
+                    move_to_GPU = True
+                    break
+
+            new_ops = None
+            outputs = []
+            # Apply the lifter
+            for lopt in (self.local_optimizers_map.get(node.op, []) +
+                         self.local_optimizers_map.get(type(node.op), []) +
+                         self.local_optimizers_all):
+                if move_to_GPU:
+                    t_opt = time.time()
+                    new_ops = lopt.transform(node.op, context_name,
+                                             [mapping[i] for i in node.inputs],
+                                             node.outputs)
+                    t_opt2 = time.time()
+                    time_opts[lopt] += t_opt2 - t_opt
+
+                    if new_ops:
+                        process_count[lopt] += 1
+                        break
+            if not new_ops:
+                newnode = node.clone_with_new_inputs([mapping.get(i)
```

Comment:
And check if the new node is not same as the old node for verifying that it has been transformed ?